### PR TITLE
Label PRs based on the labels of the associated issue

### LINKED ISSUE #17602: Create an automation to label PRs based on the associated issue

### PR DESCRIPTION
**Pull request by pedronfigueiredo**
## Explanation

We search on the body of our PR for the issue it closes (as per [our PR template](https://github.com/MetaMask/metamask-extension/blob/develop/.github/PULL_REQUEST_TEMPLATE.md?plain=1#L11)).

A one-to-one link can only be established if the PR body contains one issue number. Otherwise, we use the branch name to establish the connection between the Issue and the Pull Request. For that to work, we nudge developers to do that via the (opt-in) pre-commit hook and enforce it in the CI job.

Syncing the labels is unidirectional and non-destructive. PRs get the labels from the associated issue, but an issue does not ever get new labels automatically. Preexisting PR labels that are not on the issues persist.

No labels are copied if developers do not include an issue number on the PR body and the branch name does not adhere to the convention.

Closes #17602 

 ## How to test

- Remove PR labels
- Clear assignees for the PR
- Assign the PR to me
- The relevant job will trigger and add the issue labels automatically

### CODE REVIEW

**Comment by brad-decker** [comment]
Does the branch naming automatically link issue to PR in zenhub?

**Comment by pedronfigueiredo** [comment]
> Does the branch naming automatically link issue to PR in zenhub?

@brad.decker no, but it's used on the CI that runs on the PR as a means to know the related issue. See this line: https://github.com/MetaMask/metamask-extension/pull/17603#discussion_r1097947301

**Comment by brad-decker** [comment]
tangent: could we figure out how we can associate the pr to the issue automatically (either as part of this or a follow up) that would be a *HUGE* bonus and make PMs happy for ever.

**Review by brad-decker — DISMISSED** [review]

File: .github/workflows/label-prs.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  validate-branch-name:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Validate branch name
+        env:
+          HEAD_REF: ${{ github.event.pull_request.head.ref }}
+        uses: actions/github-script@v6
+        with:
+          script: |
+            const BRANCH_REGEX = /^main|develop|(ci|chore|docs|feat|fix|perf|refactor|revert|style)\/\d*(?:[-](?![-])\w*)*$/;
```
Comment (DDDDDanica) [inline .github/workflows/label-prs.yml:17]: would be nice to have it in readMe or developer guide for creating the branch, wdyt?

File: .github/workflows/label-prs.yml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      issues: write
+      pull-requests: write
+    steps:
+      - name: Label PR
+        env:
+          HEAD_REF: ${{ github.event.pull_request.head.ref }}
+        uses: actions/github-script@v6
+        with:
+          script: |
+            if (process.env.HEAD_REF === 'main' || process.env.HEAD_REF === 'develop') {
```
Comment (mcmire) [inline .github/workflows/label-prs.yml:43]: It might be better if we could extract this script to a separate file in `.circleci/scripts`. Then it'd get proper syntax highlighting, it could be linted, etc.

File: .validate-branch-namerc.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const BRANCH_REGEX =
```
Comment (mcmire) [inline .validate-branch-namerc.js:1]: Should this be a dotfile? Also typically files that end in `rc` are config files, but this is a script. Wondering if it makes sense to move this to the `development` directory and call this `validate-branch-name.js`.

File: README.md [hunk 4]
```diff
@@ -15,21 +15,27 @@ To learn how to contribute to the MetaMask project itself, visit our [Internal D
 ## Building locally
 
 - Install [Node.js](https://nodejs.org) version 16
-    - If you are using [nvm](https://github.com/nvm-sh/nvm#installing-and-updating) (recommended) running `nvm use` will automatically choose the right node version for you.
```
Comment (mcmire) [inline README.md:18]: Were these changes intentional? If Prettier necessitated these changes, then I'm wondering why we haven't had an issue up to now. If not, then perhaps extract these to a new PR?

File: .github/workflows/label-prs.yml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+on:
+  pull_request:
+    types: [assigned, opened, edited, synchronize, reopened]
+
+jobs:
+  validate-branch-name:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Validate branch name
```
Comment (mcmire) [inline .github/workflows/label-prs.yml:11]: I'm curious whether it makes sense to run this check in CI. At this point, the developer has created a branch and pushed it up. So if the name of that branch is not in the expected format and they see this check failing, the developer would have to close the PR, rename the branch locally, then make a new PR against that branch. I feel like that might lead to a painful process for people who don't know about this (new hires and contributors).

Is it absolutely necessary that PRs get assigned the correct labels? I'm wondering if it would be better to remove this job and update the job below this so that it bails early if it can't read the issue number from the name of the branch.

File: .husky/pre-commit [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+yarn validate-branch-name
```
Comment (mcmire) [inline .husky/pre-commit:4]: What do you think about running this on pre-push rather than pre-commit? Because the labeling occurs in a CircleCI job, it seems that we only really care about branches as soon as they appear on GitHub. It would also be less intrusive for people who commit frequently or employ rebasing frequently.

**Comment by pedronfigueiredo** [comment]
> could we figure out how we can associate the pr to the issue automatically (either as part of this or a follow up) that would be a HUGE bonus and make PMs happy for ever.

> Is it absolutely necessary that PRs get assigned the correct labels? I'm wondering if it would be better to remove this particular job and update the job below this so that it bails early if it can't read the issue number from the name of the branch.

Great points, @brad-decker @mcmire. I did a bit more reading and:

GitHub [provides linking][linking-pr], but we can't rely on people to include the right format even though it is in [our PR template][pr-template]. However, I decided to add mentions in the PR text as an additional method for getting the issue number.

In summary, the order of operations now is as follows:
- if only one issue is mentioned in the PR body, get the labels off that issue.
- if no issue is mentioned, or more than one is
  - Check if the branch name has the expected format (exclude main, develop, and rc branches)
    - if it is, get the issue number off the branch name and copy the labels.
    - if it isn't, bail with no error, so the job still passes. No labels are copied.

[linking-pr]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
[pr-template]: https://github.com/MetaMask/metamask-extension/blob/develop/.github/PULL_REQUEST_TEMPLATE.md?plain=1#L11

File: .circleci/scripts/label-prs.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
+async function runLabelPRsAutomation(context, github) {
```
Comment (brad-decker) [inline .circleci/scripts/label-prs.js:1]: Only things that run in circleci should be in this folder probably, might make sense to add it to the .github folder as it is ran by the github actions.

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: .validate-branch-namerc.js (hunk 3 above)
Comment (mcmire) [inline .validate-branch-namerc.js:1]: Also, going along with the other suggestion above, could this also be a TypeScript file? If it's too much trouble because it requires monkeying with the TypeScript/ESLint config, no problem, but I thought I'd mention it.

File: .github/scripts/label-prs.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    process.exit(1);
+  }
+
+  const prBody = context.payload.pull_request?.body || '';
+  return prBody;
+}
+
+function bailIfIsBranchNameInvalid(branchName: string): void {
+  const BRANCH_REGEX =
+    /^(main|develop|(ci|chore|docs|feat|feature|fix|perf|refactor|revert|style)\/\d*(?:[-](?![-])\w*)*|Version-v\d+\.\d+\.\d+)$/;
```
Comment (DDDDDanica) [inline .github/scripts/label-prs.ts:69]: we can also optional chain `beta` after version one, wdyt?

**Review by DDDDDanica — APPROVED** [review]

**Review by albertolive — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]
